File: baekjoon/1389.py
from collections import deque
n, m = map(int, input().split())
graph = [[] for _ in range(n+1)]
for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

def bfs(now):
    q = deque()
    q.append(now)
    used = [now]
    cnt = [0]*(n+1)
    while q:
        now = q.popleft()
        for i in graph[now]:
            if i not in used:
                q.append(i)
                used.append(i)
                cnt[i] = cnt[now] + 1
    return sum(cnt)

result = []
for i in range(1, n+1):
    result.append(bfs(i))

print(result.index(min(result))+1)

File: baekjoon/1389.js
// https://www.acmicpc.net/problem/1389
// 케빈 베이컨의 6단계 법칙
// 플로이드 워셜

const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const [n, m] = input.shift().split(" ").map(Number);
const route = input.map((v) => v.split(" ").map(Number));

function solution() {
  const INF = 21e8;
  let dist = Array.from(Array(n), () => Array(n).fill(INF));
  route.forEach(([start, end]) => {
    dist[start - 1][end - 1] = 1;
    dist[end - 1][start - 1] = 1;
  });

  for (let d = 0; d < n; d++) {
    for (let i = 0; i < n; i++) {
      dist[i][i] = 0;
      for (let j = 0; j < n; j++) {
        dist[i][j] = Math.min(dist[i][j], dist[i][d] + dist[d][j]);
      }
    }
  }

  let minValue = INF;
  let minIndex = -1;
  dist.forEach((col, index) => {
    if (col.reduce((sum, current) => sum + current) < minValue) {
      minValue = col.reduce((sum, current) => sum + current);
      minIndex = index;
    }
  });
  console.log(minIndex + 1);
}
solution();
